### ISSUE TITLE
A1: Scaled Dot-Product Attention, Multi-Head Attention, and BPE Tokenizer

### ISSUE THREAD

**Issue by anantsimran** [anantsimran/dawnshard#15]
## Overview

**Spine:** encoder-only text classifier on AG News
**Dataset:** AG News sampled (~20k train / 2k val)
**Paper components:** scaled dot-product attention · multi-head attention · BPE tokenizer
**Target hardware:** Mac MPS

Implement attention from scratch and a BPE tokenizer from scratch, then use them to build an encoder-only topic classifier. The BPE object trained here is reused in A2–A4, so treat it as a real interface.

---

## Problem statement

1. Implement `scaled_dot_product_attention(q, k, v, mask=None) -> (out, attn)`.
2. Implement `MultiHeadAttention(d_model, n_heads)` as an `nn.Module` returning `(out, attn)`.
3. Implement `BPE.train` and `BPE.encode` in `a1/helpers.py` (merge loop and greedy encoding are left blank).
4. Build `Classifier` = embedding → 1–2 encoder-style attention blocks → mask-aware mean pool → linear to 4 logits.
5. Train on sampled AG News, report validation accuracy, and produce per-head attention heatmaps for 3 example sentences.

**Conceptual question:** Before training, predict which token positions a single head should attend to for topic classification. After training, plot the heads and report whether any head specialized as predicted.

---

## Shape reference

```
ids        (B, L)            long
embed      (B, L, d)
Q, K, V    (B, L, d) -> reshape to (B, h, L, d_k),  d_k = d / h
scores     (B, h, L, L)      = Q @ K^T / sqrt(d_k)
attn       (B, h, L, L)      = softmax(scores + additive_mask, dim=-1)
context    (B, h, L, d_k)    = attn @ V  -> reshape back to (B, L, d)
pooled     (B, d)            mask-aware mean over L
logits     (B, 4)
```

**Why `sqrt(d_k)`:** dot products of `d_k`-dim random vectors have std ≈ `sqrt(d_k)`. Without the divisor, scores saturate softmax to near-one-hot and gradients vanish. At `d_k=64`, raw scores reach ±8; after dividing by 8 they sit near ±1 where softmax gradients are healthy.

---

## Allowed / banned

- **Banned:** `torch.nn.functional.scaled_dot_product_attention`, `nn.MultiheadAttention`, `nn.TransformerEncoderLayer`, `nn.Transformer`
- **Allowed:** `nn.Linear`, `nn.Embedding`, `nn.LayerNorm`, `nn.Dropout`, `softmax`, standard tensor ops

---

## Deliverables

- [x] `a1/model.py` — `scaled_dot_product_attention`, `MultiHeadAttention`, `Classifier`
- [x] `a1/helpers.py` — `BPE.train` and `BPE.encode` filled in
- [x] `a1/bpe_model.json` — trained tokenizer (committed; A2–A4 load it)
- [x] `a1/train.py` — training script printing val accuracy
- [x] `a1/attention.png` — per-head heatmaps for 3 sentences
- [ ] `a1/WRITEUP.md` — prediction vs. observed head behavior; one paragraph on `sqrt(d_k)`

---

## Grading (100 pts)

| Criterion | Pts |
|---|---|
| Attention math correct, mask applied before softmax | 25 |
| BPE trains and `decode(encode(x))` round-trips | 20 |
| From-scratch compliance (no banned ops) | 15 |
| Per-head viz + prediction-vs-observed writeup | 20 |
| Conceptual answers (`sqrt(d_k)`, head specialization) | 10 |
| Code quality: reusable interfaces, docs, function size ≤40 lines | 10 |

`pytest a1/test_a1.py` must pass — tests check softmax normalization, `sqrt(d_k)` divisor, padding-mask zeroing, MHA/classifier shapes, and BPE round-trip.

---

## Setup

```zsh
python -m venv .venv && source .venv/bin/activate
pip install torch datasets matplotlib pytest

# Verify MPS is available
python -c "from common.device import get_device; print(get_device())"
# expect: mps
```

---

### 1. `scaled_dot_product_attention`

```python
import math
import torch
import torch.nn.functional as F

def scaled_dot_product_attention(q, k, v, mask=None):
    # q, k, v: (B, h, L, d_k)
    d_k = q.size(dim=-1)
    scores = q @ k.transpose(dim0=-2, dim1=-1) / math.sqrt(d_k)  # (B, h, L, L)
    if mask is not None:
        scores = scores + mask  # additive: 0 keeps, -inf blocks
    attn = F.softmax(scores, dim=-1)
    out = attn @ v  # (B, h, L, d_k)
    return out, attn
```

### 2. `MultiHeadAttention` skeleton

```python
class MultiHeadAttention(nn.Module):
    def __init__(self, d_model: int, n_heads: int):
        super().__init__()
        assert d_model % n_heads == 0
        self.n_heads = n_heads
        self.d_k = d_model // n_heads
        self.w_q = nn.Linear(in_features=d_model, out_features=d_model)
        self.w_k = nn.Linear(in_features=d_model, out_features=d_model)
        self.w_v = nn.Linear(in_features=d_model, out_features=d_model)
        self.w_o = nn.Linear(in_features=d_model, out_features=d_model)

    def forward(self, query, key, value, mask=None):
        batch_size = query.size(dim=0)
        # project and split into heads
        def split_heads(x):
            return x.view(batch_size, -1, self.n_heads, self.d_k).transpose(dim0=1, dim1=2)
        q = split_heads(self.w_q(query))
        k = split_heads(self.w_k(key))
        v = split_heads(self.w_v(value))
        context, attn = scaled_dot_product_attention(q=q, k=k, v=v, mask=mask)
        # merge heads back
        context = context.transpose(dim0=1, dim1=2).contiguous().view(batch_size, -1, self.n_heads * self.d_k)
        return self.w_o(context), attn
```

### 3. BPE training skeleton

```python
def train(self, corpus):
    # Step 1: character-level pre-tokenization with </w> end-of-word marker
    word_freqs = Counter()
    for line in corpus:
        for word in line.strip().split():
            word_freqs[" ".join(list(word)) + " </w>"] += 1

    vocab = {word: freq for word, freq in word_freqs.items()}

    # Step 2: merge loop
    for _ in range(self.vocab_size - 256):
        pair_freqs = Counter()
        for word, freq in vocab.items():
            symbols = word.split()
            for i in range(len(symbols) - 1):
                pair_freqs[(symbols[i], symbols[i + 1])] += freq
        if not pair_freqs:
            break
        best_pair = max(pair_freqs, key=pair_freqs.get)
        self.merges.append(best_pair)
        merged = "".join(best_pair)  # e.g. ('h', 'e') -> 'he'
        vocab = {
            word.replace(" ".join(best_pair), merged): freq
            for word, freq in vocab.items()
        }

    # Step 3: build vocabulary maps (reserve 0 for <pad>)
    all_tokens = set()
    for word in vocab:
        all_tokens.update(word.split())
    self.token_to_id = {"<pad>": 0}
    for token in sorted(all_tokens):
        self.token_to_id[token] = len(self.token_to_id)
    self.id_to_token = {v: k for k, v in self.token_to_id.items()}
```

### 4. Building the padding mask

```python
# In your Classifier.forward:
from common.masks import make_padding_mask, to_additive

padding_mask = make_padding_mask(ids=ids, pad_id=pad_id)   # (B, 1, 1, L)
additive_mask = to_additive(mask=padding_mask)              # 0 / -inf
out, attn = self.mha(query=x, key=x, value=x, mask=additive_mask)

# Mask-aware mean pool
mask_float = padding_mask.squeeze(dim=1).squeeze(dim=1).unsqueeze(dim=-1)  # (B, L, 1)
pooled = (out * mask_float).sum(dim=1) / mask_float.sum(dim=1)
```

---

## Sources

- Vaswani et al., *Attention Is All You Need*, NeurIPS 2017 (§3.2.1 scaling)
- Sennrich, Haddow, Birch, *NMT of Rare Words with Subword Units*, ACL 2016
- Rush, *The Annotated Transformer*, Harvard NLP
